credential/schema/vcjsonschema.go

```diff
@@ -1,77 +1,64 @@
+func ValidateCredentialAgainstSchema(access VCJSONSchemaAccess, cred credential.VerifiableCredential) error {
+	jsonSchema, _, err := GetCredentialSchemaFromCredential(access, cred)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not marshal vc json schema's schema property")
+		return errors.Wrap(err, "error getting schema from credential")
 	}
-	maybeSchema := string(schemaBytes)
-	if err = schema.IsValidJSONSchema(maybeSchema); err != nil {
-		return nil, errors.Wrap(err, "vc json schema did not contain a valid JSON Schema")
+	if err = IsCredentialValidForJSONSchema(cred, jsonSchema); err != nil {
+		return errors.Wrap(err, "credential not valid for schema")
 	}
-	return &vcs, nil
+	return nil
 }
 
-// IsValidCredentialSchema determines if a given credential schema is compliant with the specification's
-// JSON Schema https://w3c-ccg.github.io/vc-json-schemas/v2/index.html#credential_schema_definition
-func IsValidCredentialSchema(maybeCredentialSchema string) error {
-	vcJSONSchemaSchema, err := schema.LoadSchema(schema.VerifiableCredentialJSONSchemaSchema)
+// IsCredentialValidForJSONSchema validates a credential against a schema, returning an error if it is not valid
+func IsCredentialValidForJSONSchema(cred credential.VerifiableCredential, s JSONSchema) error {
+	if !IsSupportedVCJSONSchemaType(cred.CredentialSchema.Type) {
+		return fmt.Errorf("credential schema type<%s> is not supported", cred.CredentialSchema.Type)
+	}
+	if !IsSupportedJSONSchemaVersion(s.Schema()) {
+		return fmt.Errorf("schema version<%s> is not supported", s.Schema())
+	}
+	schemaBytes, err := json.Marshal(s)
 	if err != nil {
-		return errors.Wrap(err, "could not get known schema for VC JSON Schema")
+		return errors.Wrap(err, "error marshalling schema")
 	}
-
-	if err = schema.IsValidAgainstJSONSchema(maybeCredentialSchema, vcJSONSchemaSchema); err != nil {
-		return errors.Wrap(err, "credential schema did not validate")
+	credBytes, err := json.Marshal(cred)
+	if err != nil {
+		return errors.Wrap(err, "error marshalling credential")
 	}
-
-	if _, err = StringToVCJSONCredentialSchema(maybeCredentialSchema); err != nil {
-		return errors.Wrap(err, "credential schema not valid")
+	if err = schema.IsValidAgainstJSONSchema(string(credBytes), string(schemaBytes)); err != nil {
+		return errors.Wrap(err, "credential not valid for schema")
 	}
-
 	return nil
 }
 
-func IsCredentialValidForVCJSONSchema(cred credential.VerifiableCredential, vcJSONSchema VCJSONSchema) error {
-	schemaBytes, err := json.Marshal(vcJSONSchema.Schema)
-	if err != nil {
-		return err
+// GetCredentialSchemaFromCredential returns the credential schema and type for a given credential given
+// a credential schema access, which is used to retrieve the schema
+func GetCredentialSchemaFromCredential(access VCJSONSchemaAccess, cred credential.VerifiableCredential) (JSONSchema, VCJSONSchemaType, error) {
```

Comment:
Consider making this a receiver method of `VCJSONSchemaAccess`

Reply:
good idea

Reply:
no receiver on interfaces 😢

Reply:
Another method of the interface then?

Reply:
that would lead to re-implementing this logic for each access, when it could be defined once